crates/core/src/error.rs

```diff
@@ -1,810 +1,451 @@
-#[derive(Error, Debug, Display)]
-pub enum ConfigFileErrorKind {
-    /// config version not supported!
-    ConfigVersionNotSupported,
-    /// Parsing Polynomial in config failed: `{0:?}`
-    ParsingFailedForPolynomial(#[from] ParseIntError),
-}
+    /// Severity of the error.
+    severity: Option<Severity>,
 
-/// [`KeyFileErrorKind`] describes the errors that can be returned for `KeyFile`s
-#[derive(Error, Debug, Display)]
-pub enum KeyFileErrorKind {
-    /// no suitable key found!
-    NoSuitableKeyFound,
-    /// listing KeyFiles failed
-    ListingKeyFilesFailed,
-    /// couldn't get KeyFile from backend
-    CouldNotGetKeyFileFromBackend,
-    /// serde_json couldn't deserialize the data: `{0:?}`
-    DeserializingFromSliceFailed(serde_json::Error),
-    /// couldn't encrypt data: `{0:?}`
-    CouldNotEncryptData(#[from] CryptoErrorKind),
-    /// serde_json couldn't serialize the data into a JSON byte vector: `{0:?}`
-    CouldNotSerializeAsJsonByteVector(serde_json::Error),
-    /// conversion from `u32` to `u8` failed: `{0:?}`
-    ConversionFromU32ToU8Failed(TryFromIntError),
-    /// output length is invalid: `{0:?}`
-    OutputLengthInvalid(scrypt::errors::InvalidOutputLen),
-    /// invalid scrypt parameters: `{0:?}`
-    InvalidSCryptParameters(scrypt::errors::InvalidParams),
-}
+    /// The status of the error.
+    status: Option<Status>,
 
-/// [`PackFileErrorKind`] describes the errors that can be returned for `PackFile`s
-#[derive(Error, Debug, Display)]
-pub enum PackFileErrorKind {
-    /// Failed reading binary representation of the pack header: `{0:?}`
-    ReadingBinaryRepresentationFailed(binrw::Error),
-    /// Failed writing binary representation of the pack header: `{0:?}`
-    WritingBinaryRepresentationFailed(binrw::Error),
-    /// Read header length is too large! Length: `{size_real}`, file size: `{pack_size}`
-    HeaderLengthTooLarge { size_real: u32, pack_size: u32 },
-    /// Read header length doesn't match header contents! Length: `{size_real}`, computed: `{size_computed}`
-    HeaderLengthDoesNotMatchHeaderContents { size_real: u32, size_computed: u32 },
-    /// pack size computed from header doesn't match real pack isch! Computed: `{size_computed}`, real: `{size_real}`
-    HeaderPackSizeComputedDoesNotMatchRealPackFile { size_real: u32, size_computed: u32 },
-    /// partially reading the pack header from packfile failed: `{0:?}`
-    ListingKeyFilesFailed(#[from] BackendAccessErrorKind),
-    /// decrypting from binary failed
-    BinaryDecryptionFailed,
-    /// Partial read of PackFile failed
-    PartialReadOfPackfileFailed,
-    /// writing Bytes failed
-    WritingBytesFailed,
-    /// `{0}`
-    #[error(transparent)]
-    PackDecryptionFailed(#[from] CryptBackendErrorKind),
+    /// Backtrace of the error.
+    ///
+    // Need to use option, otherwise thiserror will not be able to derive the Error trait.
+    backtrace: Option<Backtrace>,
```

Comment:
sad it's `Option`al but as we deal internally with it, `rustic_core` developers can make sure it's never put as `None`.

Reply:
Yeah, but sometimes, we don't have a source, e.g. when coming from `ok_or_else`